rfcs/0015-const-generics.md

```diff
@@ -0,0 +1,299 @@
+ [0; N + 1]
+}
+```
+
+This poses the challenge of having an expression tree inside the type system. Currently `sway` already 
+has three expression trees: `Expr`, `ExpressionKind` and `TyExpressionVariant`. This demands a new one,
+given that the first two allow much more complex expressions than what `const generics` wants to support; 
+and  last one is only create after some `TypeInfo` already exist, and thus cannot be used.
+
+At the same time, the parser should be able to parse any expression and return a friendly error that such expression,
+is not supported. 
+
+So, in case of a unsupported expression the parser will parser the `const generic` expression as it does for normal `Expr`and will lower it to the type system expression enum, but in the place of the unsupported expression will return a `TypeInfo::ErrorRecovery`.
+
+These expressions will also increase the complexity of all types related algorithms such as:
+
+1. Unification
+2. PartialEq
+3. Hash
+
+In the simplest case, it is very clear how to unify `TypeInfo::Array(..., Length::Literal(1))` and `TypeInfo::Array(..., Length::Expression("N"))`.
+But more complex cases such as `TypeInfo::Array(..., Length::Expression("N"))` and `TypeInfo::Array(..., Length::Expression("N + 1"))`, is not clear
+if these types are unifiable, equal or simply different.
+
+## Method call search algorithm
+
+When a method is called, the algorithm that searches which method is called uses the method `TraitMap::get_impls`.
+Currently, this method does an `O(1)` search to find all methods applicable to a type. For example:
+
+```sway
+impl [u64; 1] {
+ fn len_for_size_one(&self) { ... }
+}
+```
+
+would create a map with something like
+
+```
+Placeholder -> [...]
+...
+[u64; 1] -> [..., len_for_size_one,...]
+...
+```
+
+The algorithms first create a `TypeRootFilter`, which is very similar to `TypeInfo`. And uses this `enum` to search the hash table.
+After that, it "generalizes" the filter and searches for `TypeRootFilter::Placeholder`.
+
+To fully support `const generics` and `const value specialization`, the compiler will now keep generalizing the
+searched type until it hits `Placeholder`. For example, searching for `[u64; 1]` will actually search for:
+
+1. [u64; 1];
+1. [u64; Placeholder];
+1. Placeholder;
+
+The initial implementation will do this generalization only for `const generics`, but it also makes sense to 
+generalize this with other types such as `Vec<u64>`.
+
+1. Vec\<u64>;
+1. Vec\<Placeholder>;
+1. Placeholder;
+
+This gets more complex as the number of `generics` and `const generics` increases. For example:
+
+```sway
+struct VecWithSmallVecOptimization<T, const N: u64> { ... }
+```
+
+Searching for this type would search:
+
+1. VecWithSmallVecOptimization\<u64, 1>
+1. VecWithSmallVecOptimization\<Placeholder, 1>
+1. VecWithSmallVecOptimization\<u64, Placeholder>
+1. VecWithSmallVecOptimization\<Placeholder, Placeholder>
+1. Placeholder
+
```

Comment:
Perhaps to explicitly mention that the insertion into trait map must already search for conflicting implementations in regard to const generic arguments. 
```suggestion
When inserting into the `TraitMap` the search for duplicated definitions will need to be extended for const generic arguments. E.g.:
```sway
struct S<const N:u8> { }

impl<const N:u8> S<N> {
    fn a(&self) { }
}

impl S<0> {
    fn a(&self) { } // <<<--- ERROR: Conflicting definition.
    fn b(&self) { } // <<<--- OK.
}

trait Trait { }

impl<const N:u64> Trait for [u8;N] { }

impl Trait for [u8;2] { } <<<--- ERROR: Conflicting implementation.
``` 
```

Reply:
> impl S<0> {

This specialization will not be supported.